File: Old Clean/AonWeekly_Clean.r

library(dplyr)
library(DBI)
library(RODBC)
library(readr)
library(readxl)

db_conn <- odbcConnect("ILS")

#################################################################################################################

#FILE NEEDS TO BE CLEANED BEFORE USE
table <- read_excel("S:/Touchstone/Catrader/Boston/Database/CatBond/Shiny/Aon_Weekly/Aon20231229Clean.xlsx", sheet = 'RLS', col_names=T, skip=1) %>%
		data.frame()

#DATE NEEDS TO BE MANUALLY CHANGED
QDate <- rep(as.Date("2023-12-29"), times=nrow(table))
table <-cbind(QDate, table)

table <- table %>%
		select(-4,-13,-14,-15)

ColNames <- sqlColumns(db_conn, "Aon") %>% 
			select('COLUMN_NAME')

colnames(table) <- ColNames$COLUMN_NAME

table$LongTermAsk[is.na(table$LongTermAsk)] <- 0
table$LongTermEL[is.na(table$LongTermEL)] <- 0
table$NearTermAsk[is.na(table$NearTermAsk)] <- 0
table$NearTermEL[is.na(table$NearTermEL)] <- 0
table$BidSpread[table$BidSpread == 'n/a'] <- 0
table$OfferSpread[table$OfferSpread == 'n/a'] <- 0

table$Coupon <- table$Coupon %>% parse_number() %>% as.integer()
table$BidPrice <- table$BidPrice %>% parse_number()
table$OfferPrice <- table$OfferPrice %>% parse_number()

table$Size <- formatC(c(table$Size), digits = 2, format = 'f')
table$LongTermAsk <- formatC(c(table$LongTermAsk), digits = 2, format = 'f')
table$LongTermEL <- formatC(c(table$LongTermEL), digits = 2, format = 'f')
table$NearTermAsk <- formatC(c(table$NearTermAsk), digits = 2, format = 'f')
table$NearTermEL <- formatC(c(table$NearTermEL), digits = 2, format = 'f')
table$Coupon <- formatC(c(table$Coupon), digits = 2, format = 'f')
table$BidPrice <- formatC(c(table$BidPrice), digits = 2, format = 'f')
table$OfferPrice <- formatC(c(table$OfferPrice), digits = 2, format = 'f')

sqlSave(db_conn, table, tablename="Aon", rownames=F, append=T, fast=F)

##################################################################################################################################

odbcClose(db_conn)

##################################################################################################################################

File: New Clean/AonWeekly_Clean.py
import pandas as pd
import pyodbc
from datetime import datetime

# Replace with your actual connection details
conn = pyodbc.connect('DSN=ILS;Trusted_Connection=yes;')

# FILE NEEDS TO BE CLEANED BEFORE USE
file_path = "S:/Touchstone/Catrader/Boston/Database/CatBond/Shiny/Aon_Weekly/Aon20231229Clean.xlsx"
table = pd.read_excel(file_path, sheet_name='RLS', skiprows=1)

# DATE NEEDS TO BE MANUALLY CHANGED
table['QDate'] = pd.Timestamp('2023-12-29')

# Dropping specific columns
table = table.drop(table.columns[[3, 12, 13, 14]], axis=1)

# Getting column names from the SQL table
cursor = conn.cursor()
cursor.execute("SELECT COLUMN_NAME FROM INFORMATION_SCHEMA.COLUMNS WHERE TABLE_NAME = 'Aon'")
col_names = [row[0] for row in cursor.fetchall()]
table.columns = col_names

# Handling NaN and 'n/a' values
table['LongTermAsk'].fillna(0, inplace=True)
table['LongTermEL'].fillna(0, inplace=True)
table['NearTermAsk'].fillna(0, inplace=True)
table['NearTermEL'].fillna(0, inplace=True)
table['BidSpread'] = table['BidSpread'].replace('n/a', 0)
table['OfferSpread'] = table['OfferSpread'].replace('n/a', 0)

# Parsing and formatting numbers
table['Coupon'] = pd.to_numeric(table['Coupon'], errors='coerce').fillna(0).astype(int)
table['BidPrice'] = pd.to_numeric(table['BidPrice'], errors='coerce').fillna(0)
table['OfferPrice'] = pd.to_numeric(table['OfferPrice'], errors='coerce').fillna(0)

for column in ['Size', 'LongTermAsk', 'LongTermEL', 'NearTermAsk', 'NearTermEL', 'Coupon', 'BidPrice', 'OfferPrice']:
    table[column] = table[column].apply(lambda x: format(x, '.2f'))

# Saving to SQL database
for index, row in table.iterrows():
    cursor.execute("INSERT INTO Aon (columns) VALUES (values)", tuple(row))
    conn.commit()

# Closing the database connection
conn.close()
